proposals/000-decorator-api.md

```diff
@@ -0,0 +1,322 @@
+
+
+## Developer Guide
+
+### Terms
+
+* **Decoration**: A single user interface element drawn on a publication.
+* **Decoration group**: A set of decorations applied together. Each group is named by the developer, usually according to the feature the decorations belong to (i.e. `search`, `highlights`, `tts`).
+* **Decoration style**: Determines the look and feel of a decoration once rendered by a Navigator. For example `highlight`, `underline` or `margin icon`. It is media type agnostic.
+* **Decoration template**: Translates a decoration and its style into concrete rendering instructions for the Navigator. Template implementations depend on the resource media type and the particular Navigator used.
+
+### Creating a New Decoration
+
+The `Decoration` object associates a location (`Locator`) in a publication with a *decoration style* to render. By relying on the style to indicate its look and feel, `Decoration` avoids any specific rendering instructions. This allows to reuse the same decorating code for any kind of publication or different Navigators.
+
+Each `Decoration` object has an identifier which **must** be unique in the decoration's group. This is necessary to compute efficiently changes when applying updates. A database ID is a perfect candidate.
+
+```swift
+Decoration(
+    id: "12",
+    locator: Locator(...),
+    style: Decoration.Style.MarginIcon(image: "bookmark")
+)
+```
+
+Note that a single entity might be rendered using several `Decoration` objects. For example, a highlight with a text annotation needs two `Decoration` objects: one for the actual highlight and one to draw the "text annotation" icon in the page margin.
+
+### Applying Decorations to a Navigator
+
+A reading app never explicitly asks the Navigator to render a decoration. Instead, the app declares the current state of the decorations for a given *decoration group* and the Navigator will decide when to render each decoration. Navigators use a diff algorithm to determine the actual changes since the last applied state to run efficient rendering instructions.
+
+```swift
+// First, fetch highlight models from your database.
+let highlights = ...
+
+// Then, create one new decoration per highlight.
+let decorations = highlights.map { highlight in
+    Decoration(
+        id: highlight.id,
+        locator: highlight.locator,
+        style: Decoration.Style.Highlight(tint: highlight.color)
+    )
+}
+
+// Finally, apply the decorations in a group named "user-highlights".
+navigator.apply(decorations: decorations, in: "user-highlights")
+```
+
+### Handling User Clicks on Decorations
+
+To allow user clicks/taps on a decoration, you must first check that the `activate` interaction is available for the decoration style:
+
+```swift
+navigator.supportsDecorationInteraction(.activate, forStyle: Decoration.Style.Highlight)
+```
+
+Then, enable the `activate` interaction when creating each `Decoration` object.
```

Comment:
What's the rationale behind this? It looks a bit overkill. I would rather notify observers of any interaction, and let them decide whether they want to do something.

Reply:
If we have highlights in mind it seems a bit redundant, but I'm not convinced that decorations should be always "activable".  For example, if you use a decoration to render a "hand drawn note" over a PDF page. Having this activable would make it take over the links or other actually interactive elements on the page that are below the drawing. Having no enabled interactions makes it a "passthrough" for gestures and touches.

Of course this is highly dependent on the media type, in some cases it might not be an issue at all.

Reply:
I think activability depends on the group rather than on the specific decoration-instance though. If I remember correctly, you mentioned that the decoration template should specify which zone would be activable, it might be enough.

Reply:
We might also require callbacks to return booleans saying whether they have handled the interaction, as it is quite common done. It would delegate to apps the choice of the decoration to activate in case of overlapping.

Reply:
> I think activability depends on the group rather than on the specific decoration-instance though.

Yeah I think that makes sense, and registering a listener for an interaction in a group would implicitly enable the gesture on these decorations.

> We might also require callbacks to return booleans saying whether they have handled the interaction, as it is quite common done. It would delegate to apps the choice of the decoration to activate in case of overlapping.

Are you talking about the case where we have a decoration on top of another one in the same group, but the app wants to handle the gesture on the one below? And not about the case where a highlight is above a link in the publication, for example?

For a link inside the publication, personally I think the publication content should take precedence over any Decoration for interactions. This is how it's handled in the Swift version and also Desktop. Basically the decoration interaction is triggered only if the publication itself didn't handle the touch event.

Reply:
> For a link inside the publication, personally I think the publication content should take precedence over any Decoration for interactions. This is how it's handled in the Swift version and also Desktop. Basically the decoration interaction is triggered only if the publication itself didn't handle the touch event.

Makes sense. I was talking about the case where we have a decoration on top of another one, in the same group or not. With your proposal, I believe there are two choices to handle those cases:
- letting the navigator arbitrarily decide which decoration should be activated, depending on the activability of each one
- activating all activable decorations at once, that is calling listeners one after another
Actually I'm not even sure what behaviour is desirable.
-

Reply:
> activating all activable decorations at once, that is calling listeners one after another

I can't think of use case using that, so I think it's fine if the API doesn't support this at least in its first version.

> letting the navigator arbitrarily decide which decoration should be activated, depending on the activability of each one

Actually this is not specified enough in this proposal. But basically I'm following the list of decorations provided by the app, the later ones taking precedence over previous ones. This is also useful to know how to draw overlapping decorations. I think that's enough to disambiguate and letting the control to the app. I'm not sure going over all the overlapping decorations is really useful in this case (with a "handled" return value), especially since the app wouldn't know what other decorations are conflicting in this case.

An alternative would be to send all decorations that are at the interaction point as a list instead of one by one. Not sure this is really useful but that could be a compromise.

Reply:
Ok, let's drop it.